lua/mini/ai.lua

```diff
@@ -1562,24 +1577,24 @@ H.get_matched_ranges_builtin = function(captures)
   -- Get parser (LanguageTree) at cursor (important for injected languages)
   local pos = vim.api.nvim_win_get_cursor(0)
   local lang_tree = parser:language_for_range({ pos[1] - 1, pos[2], pos[1] - 1, pos[2] })
-  local lang = lang_tree:lang()
-
-  -- Get query file depending on the local language
-  local query = vim.treesitter.query.get(lang, 'textobjects')
-  if query == nil then H.error_treesitter('query', lang) end
-
-  -- Compute ranges of matched captures
-  local capture_is_requested = vim.tbl_map(function(c) return vim.tbl_contains(captures, '@' .. c) end, query.captures)
 
+  local missing_query_langs = {}
   local res = {}
-  for _, tree in ipairs(lang_tree:trees()) do
-    -- TODO: Remove `opts.all`after compatibility with Neovim=0.10 is dropped
-    for _, match, metadata in query:iter_matches(tree:root(), buf_id, nil, nil, { all = true }) do
-      for capture_id, nodes in pairs(match) do
-        local mt = metadata[capture_id]
-        if capture_is_requested[capture_id] then table.insert(res, H.get_nodes_range_builtin(nodes, buf_id, mt)) end
-      end
+  -- Recursively query parent LanguageTree as fallback (important for injected languages)
+  while vim.tbl_isempty(res) and lang_tree ~= nil do
+    local lang = lang_tree:lang()
+    -- Get query file depending on the local language
+    local query = vim.treesitter.query.get(lang, 'textobjects')
+
+    if query ~= nil then
+      H.append_ranges(res, buf_id, query, captures, lang_tree)
+    elseif query == nil and not vim.tbl_contains(missing_query_langs, lang) then
```

Comment:
`missing_query_langs` can be a map. Like `{ python = true }`, etc. It allows using plain `else` instead of `elseif` at the cost of the later `vim.tbl_keys()` call.

Also, although it comes with extra unneeded check, could you please use the following instead of `if ... else ... end`:

```lua
if query ~= nil then H.append_ranges(res, buf_id, query, captures, lang_tree) end
if query == nil then missing_queary_langs[lang] = true end
```

It feels more readable while 60% less lines.